katdal/categorical.py

```diff
@@ -236,58 +276,49 @@ def __len__(self):
+        return self._bool_per_dump([value == other for value in self._hashable_values])
 
     def __ne__(self, other):
         """Inequality comparison operator."""
-        index_truths = np.atleast_1d((self.unique_values != other)[self.indices])
-        result = np.empty(self.events[-1], dtype=np.bool)
-        for n, (start, end) in enumerate(zip(self.events[:-1], self.events[1:])):
-            result[start:end] = index_truths[n]
-        return result
+        return self._bool_per_dump([value != other for value in self._hashable_values])
 
     def __lt__(self, other):
         """Less-than comparison operator."""
-        index_truths = np.atleast_1d((self.unique_values < other)[self.indices])
-        result = np.empty(self.events[-1], dtype=np.bool)
-        for n, (start, end) in enumerate(zip(self.events[:-1], self.events[1:])):
-            result[start:end] = index_truths[n]
-        return result
+        return self._bool_per_dump([value < other for value in self._hashable_values])
 
     def __gt__(self, other):
         """Greather-than comparison operator."""
-        index_truths = np.atleast_1d((self.unique_values > other)[self.indices])
-        result = np.empty(self.events[-1], dtype=np.bool)
-        for n, (start, end) in enumerate(zip(self.events[:-1], self.events[1:])):
-            result[start:end] = index_truths[n]
-        return result
+        return self._bool_per_dump([value > other for value in self._hashable_values])
 
     def __le__(self, other):
         """Less-than-or-equal comparison operator."""
-        index_truths = np.atleast_1d((self.unique_values <= other)[self.indices])
-        result = np.empty(self.events[-1], dtype=np.bool)
-        for n, (start, end) in enumerate(zip(self.events[:-1], self.events[1:])):
-            result[start:end] = index_truths[n]
-        return result
+        return self._bool_per_dump([value <= other for value in self._hashable_values])
 
     def __ge__(self, other):
         """Greater-than-or-equal comparison operator."""
-        index_truths = np.atleast_1d((self.unique_values >= other)[self.indices])
-        result = np.empty(self.events[-1], dtype=np.bool)
-        for n, (start, end) in enumerate(zip(self.events[:-1], self.events[1:])):
-            result[start:end] = index_truths[n]
-        return result
+        return self._bool_per_dump([value >= other for value in self._hashable_values])
 
     @property
     def dtype(self):
         """Sensor value type."""
-        return self.unique_values.dtype
+        if self.unique_values:
+            # Put all values into array to find maximum length of any string type
+            test_data = np.array(self.unique_values)
+            # Beware array-valued sensors; treat their values as opaque objects
+            # This forces sensor values to be 1-D at all times (an invariant)
+            return test_data.dtype if test_data.ndim == 1 else np.object
+        else:
+            return np.object
```

Comment:
I was trying to avoid calling this on construction. The ```_hashable_values``` attribute will typically be accessed infrequently or not at all (only on comparisons, add and remove), while construction will necessarily happen.

Actually, a more relevant concern is that the ```concatenate_categorical``` function constructs a CategoricalData by assigning ```unique_values``` directly, bypassing the constructor for efficiency reasons.

Maybe the best place is within a setter on ```unique_values```.